### ISSUE TITLE
What is the jdbi-idiomatic way to express async queries/updates?

### ISSUE THREAD

**Issue by redlet** [jdbi/jdbi#1682]
If I wanted to return a CompletableFuture, for example, how would I do that? Or does jdbi support a different type (I believe I saw something about a result producer).

**Comment by stevenschlansker** [comment]
Hi, currently JDBC is synchronous only, and will be probably until Project Loom ships.

There's some work (with an unclear path forward) to introduce truly asynchronous operation: https://github.com/jdbi/jdbi/issues/1454 but since all RDBMSes I'm familiar with are thread or process-per-connection anyway, you end up doing an awful lot of chasing for very unclear gains.

My personal choice is to maintain a traditional connection-pooling synchronous DAO, and wrap any asynchronous needs in a `return CompletableFuture.supplyAsync(() -> dao.query(args), mySharedThreadPoolExecutor)` to produce a CompletableFuture.  This is a bit of unfortunate boilerplate to write but works very well in our production systems - we compose futures from our pooled db with an async Redis caching library and expose the result asynchronously through JAX-RS.

Now that I say that out loud: it actually probably wouldn't be all that hard to introduce a simple mapper for `Future<T>` that takes a configured thread pool and does the `supplyAsync` for you.  That would be a neat transitional feature, removing a lot of boilerplate, while we wait for Loom or R2DBC to lead us to the future...

**Comment by redlet** [comment]
yeah the CompletableFuture thing is what I am looking for. project loom is in a good state (i've toyed with the SDK), but as with all java features, it'll be ages before we see virtual threads in production. 

all I want is for my application's (primarily synchronous) logic to not get stalled waiting for a result set. nothing too fancy.

**Comment by leaumar** [comment]
You could try to make a generic extension to `RowMapper` that accepts an inner rowmapper and returns a new mapper to `Future<T>`? I'm curious how this will go but I don't have time for it right now.

**Comment by redlet** [comment]
well it would have to be CompletableFuture because `thenAcceptAsync` is important; Future's `get` method needs to be waited on.

**Comment by leaumar** [comment]
Yeah, sorry, I removed my remark again, I don't feel focused enough to speak confidently about async stuff

**Comment by stevenschlansker** [comment]
I'll see if I can find time to hack up a beta feature for the next release.

**Comment by redlet** [comment]
That would be brilliant. I'll keep an eye out and test as soon as it's available.
